server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -569,6 +567,62 @@ public int hashCode() {
         }
     }
 
+    public static class FailedShardUpdateTask implements ClusterStateTaskListener {
+        final ShardId shardId;
+        final String allocationId;
+        final long primaryTerm;
+        final String message;
+        @Nullable
+        final Exception failure;
+        final boolean markAsStale;
+        final ClusterStateTaskListener listener;
```

Comment:
This is quiet a lot of code duplication. I wonder if it would be better to:
* replace above direct field usage with getters
* make `FailedShardUpdateTask` consist of `FailedShardEntity` and a listener
* implement getters, equals/hascode/tostring and listener methods and delegate them to entity and listener
?